packages/type-utils/src/getConstrainedTypeAtLocation.ts

```diff
@@ -18,3 +28,52 @@ export function getConstrainedTypeAtLocation(
 
   return constrained ?? nodeType;
 }
+
+export interface ConstraintTypeInfoUnconstrained {
+  constraintType: undefined;
+  isTypeParameter: true;
+  type: ts.TypeParameter;
+}
+
+export interface ConstraintTypeInfoConstrained {
+  constraintType: ts.Type;
+  isTypeParameter: true;
+  type: ts.TypeParameter;
+}
+
+export interface ConstraintTypeInfoNonGeneric {
+  constraintType: ts.Type;
+  isTypeParameter: false;
+  type: ts.Type;
+}
+
+export type ConstraintTypeInfo =
+  | ConstraintTypeInfoConstrained
+  | ConstraintTypeInfoNonGeneric
+  | ConstraintTypeInfoUnconstrained;
+
+/**
+ * Resolves the given node's type, and returns info about whether it is a generic or ordinary type.
+ *
+ * Successor to {@link getConstrainedTypeAtLocation} due to https://github.com/typescript-eslint/typescript-eslint/issues/10438
+ */
+export function getConstraintTypeInfoAtLocation(
```

Comment:
@JoshuaKGoldberg I've just had the thought - I think this might be the wrong API to write. It might be better to just accept a type as input, like
```ts
export declare function getConstraintInfo(type: ts.Type): { 
  isTypeParameter: boolean, 
  constraint: ts.Type | undefined
};
```
This maintains a separation between _getting_ types and _analyzing_ types.

Reply:
@kirkwaiblinger just confirming, are you planning on making changes? And/or, is this a question?

Reply:
Lol. In my head, I was asking you for your opinion, yeah, but I forgot to write that out 😆 

Coming back to this, I currently lean towards: let's go with this PR in its current state, and if there's a need to extract just the constraint info part of this function that's quite easy to do in a followup when the need arises.

Feel free to offer your opinion! 😁

Reply:
If you're on board with this plan, then this PR is up to date and ready to be merged 👍

Reply:
Oh awesome! I don't have a strong preference here.

Proposal: whatever strategy you want to go with, how about marking the final creation as `@internal` so it's not part of the public API? That way we can play with it as we want.

Reply:
Made lots of changes in this arena. New utility is just internal to eslint-plugin for now, no deprecation yet. Should be much easier to review!

Reply:
I like this strategy. The union of possible constraint info points seems reasonable to me.